proc/combine/combine.go

```diff
@@ -12,139 +12,213 @@ import (
-	p.once.Do(func() {
-		for _, m := range p.parents {
-			go m.run()
+func (p *Proc) block(parent *puller) {
+	if !parent.blocked {
+		parent.blocked = true
+		p.nblocked++
+	}
+}
+
+func (p *Proc) propagateDone() error {
+	for _, parent := range p.parents {
+		if parent.blocked {
+			continue
 		}
-	})
-	if p.err != nil {
-		return nil, p.err
+	again:
+		select {
+		case <-p.queue:
+			// If a parent is waiting on the queue, we need to
+			// read the queue to avoid deadlock.  Since we
+			// are going to throw away the batch anyway, we can
+			// simply ignore which parent it is as we will hit all
+			// of them eventually as we loop over each unblocked parent.
+			goto again
+		case parent.doneCh <- struct{}{}:
+			p.block(parent)
+		case <-p.ctx.Done():
+			return p.ctx.Err()
+		}
+	}
+	// Make sure all the dones that canceled pending queue entries
+	// are clear.  Otherwise, this will block the queue on the next
+	// platoon.
+drain:
+	select {
+	case <-p.queue:
+		goto drain
+	default:
+	}
+	// Now that everyone is blocked either because they sent us an EOS,
+	// we sent them a done, or and EOS/done collided at the same time,
+	// we can unblock everything.
+	return p.unwait()
+}
+
+type puller struct {
+	zbuf.Puller
+	ctx      context.Context
+	resultCh chan proc.Result
+	doneCh   chan struct{}
+	waitCh   chan<- struct{}
+	queue    chan<- *puller
+	// used only by Proc
+	blocked bool
+}
+
+func newPuller(ctx context.Context, waitCh chan<- struct{}, parent zbuf.Puller, q chan<- *puller) *puller {
+	return &puller{
+		Puller:   proc.NewCatcher(parent),
+		ctx:      ctx,
+		resultCh: make(chan proc.Result),
+		doneCh:   make(chan struct{}),
+		waitCh:   waitCh,
+		queue:    q,
 	}
+}
+
+func (p *puller) run() {
 	for {
-		if len(p.blocked) == len(p.parents) {
-			for _, parent := range p.parents {
-				parent.resumeCh <- struct{}{}
-				delete(p.blocked, parent)
-			}
-			return nil, nil
-		}
+		batch, err := p.Pull(false)
+		p.queue <- p
 		select {
-		case res := <-p.resultCh:
-			if err := res.err; err != nil {
-				p.err = err
-				return nil, err
+		case p.resultCh <- proc.Result{batch, err}:
+			if err != nil {
+				return
 			}
-			if res.batch == nil {
-				p.blocked[res.puller] = struct{}{}
+			if batch == nil {
+				// We just sent an EOS, so we'll wait until
+				// all the other paths are done before pulling
+				// again.  We also are guaranteed here that the
+				// combiner has our EOS and knows we're done and
+				// will mark us blocked and not raise our doneCh.
+				p.wait()
+			}
+		case <-p.doneCh:
+			if batch == nil {
+				// Combiner tells us we're done but we just
+				// received an EOS from upstream, so we don't want
+				// to call Pull(true) as they would break the contract.
+				// Since the combiner thinks we're done and our parent
+				// thinks we're done, there's nothing to do.
+				// Just continue the loop and reach for the next
+				// platoon.
+				if ok := p.wait(); !ok {
```

Comment:
Nit: This pattern appears often in this change. I think we should simplify it everywhere.
```suggestion
				if !p.wait() {
```

Reply:
The reason I like my pattern is that the wait return is that things aren't ok vs the wait is over.  There are two things going on.  But I will change it to your pattern since its shorter and you can see two things at once!